src/RlnBase.sol

```diff
@@ -97,37 +108,64 @@ abstract contract RlnBase {
         _;
     }
 
-    constructor(uint256 membershipDeposit, uint256 depth, address _verifier) {
+    modifier onlyValidUserMessageLimit(uint256 messageLimit) {
+        if (messageLimit > MAX_MESSAGE_LIMIT) revert InvalidUserMessageLimit(messageLimit);
+        if (messageLimit == 0) revert InvalidUserMessageLimit(messageLimit);
+        _;
+    }
+
+    constructor(uint256 membershipDeposit, uint256 depth, uint256 maxMessageLimit, address _verifier) {
         MEMBERSHIP_DEPOSIT = membershipDeposit;
+        MAX_MESSAGE_LIMIT = maxMessageLimit;
         DEPTH = depth;
         SET_SIZE = 1 << depth;
         verifier = IVerifier(_verifier);
         deployedBlockNumber = uint32(block.number);
     }
 
+    /// Returns the deposit amount required to register as a member
+    /// @param userMessageLimit The message limit of the member
+    /// TODO: update this function as per tokenomics design
+    function getDepositAmount(uint256 userMessageLimit) public view returns (uint256) {
+        return userMessageLimit * MEMBERSHIP_DEPOSIT;
+    }
+
     /// Allows a user to register as a member
     /// @param idCommitment The idCommitment of the member
-    function register(uint256 idCommitment) external payable virtual onlyValidIdCommitment(idCommitment) {
-        if (msg.value != MEMBERSHIP_DEPOSIT) {
+    /// @param userMessageLimit The message limit of the member
+    function register(
+        uint256 idCommitment,
+        uint256 userMessageLimit
+    )
+        external
+        payable
+        virtual
+        onlyValidIdCommitment(idCommitment)
+        onlyValidUserMessageLimit(userMessageLimit)
+    {
+        uint256 requiredDeposit = userMessageLimit * MEMBERSHIP_DEPOSIT;
```

Comment:
Isn't this same as `getDepositAmount()` ?
Would it make sense to reuse this here?

Reply:
err, missed that - addressed in dc98215 thanks!